File: 백준/Silver/15649. N과 M （1）/N과 M （1）.java
import java.io.*;
import java.util.StringTokenizer;

public class Main {

    static int N, M;
    static boolean[] visited;
    static int[] arr;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        StringTokenizer st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());

        visited = new boolean[N + 1];
        arr = new int[M];
        backtrace(0);
    }

    private static void backtrace(int depth) {
        if (depth == M) {
            // 수열 출력
            for (int i = 0; i < arr.length; i++) {
                System.out.print(arr[i] + " ");
            }
            System.out.println();
            return;
        }

        for (int i = 1; i <= N; i++) {
            if (!visited[i]) {
                visited[i] = true;
                arr[depth] = i;
                backtrace(depth + 1);
                visited[i] = false;
            }
        }

    }


}

File: 백준/Silver/15649. N과 M （1）/N과 M （1）.py
import sys
sys.setrecursionlimit(10**6)

input = sys.stdin.readline

N, M = map(int, input().split())

visited = [False] * (N + 1)
arr = [0] * M


def backtrace(depth):
    if depth == M:
        print(*arr)
        return

    for i in range(1, N + 1):
        if not visited[i]:
            visited[i] = True
            arr[depth] = i
            backtrace(depth + 1)
            visited[i] = False


backtrace(0)